jest-diff: Add options for colors and symbols (#8841)
facebook/jest: packages/jest-diff/README.md

# jest-diff

Display differences clearly so people can review changes confidently.

The default export serializes JavaScript **values** and compares them line-by-line.

Two named exports compare **strings** character-by-character:

> - `diffStringsAligned` returns a string which includes comparison lines.

Reviewer: It took me quite a while (even after reading the whole readme) to understand the difference between those two. I still do not quite understand the naming, which I think was part of what confused me because I associated things like the beginning of lines with "aligned".
Would you say that unified diff vs split diff (like for example in GitHub) is a good analogy for those functions and could be used for naming?

Reviewer 2: Agreed on being a bit confused about what `aligned` means. Unified vs split makes more sense to me (that's also the naming bitbucket uses, so I think it's pretty standard)

Author: Brilliant analogy!

* so-called aligned is like unified diff
* so-called unaligned is used in Jest to display one-line strings in normal compact report:

```text
Expected: "change from"
Received: "change to"
```

For multiline strings, a split like GitHub would consist of array of string tuples:

```js
[
  [a0, b0],
  [a1, b1],
  // and so on
]
```

for corresponding changed chunks. For deleted `b` is empty, for inserted `a` is empty.

I am curious, can y’all think of a use case in third-party dependents?

Author: * I have renamed `diffStringsAligned` and `diffStringsUnified`
* I think that I will replace `diffStringsUnaligned` with `diffStringsRaw` that returns an array of `Diff` objects, so that it will be reusable to the max

- `diffStringsUnaligned` returns an array of 2 strings.

## Installation

To add this package as a dependency of a project, run either of the following commands:

- `npm install jest-diff`
- `yarn add jest-diff`

## Usage of default export

Given values and optional options, `diffLines(a, b, options?)` does the following:

- **serialize** the values as strings using the `pretty-format` package
- **compare** the strings line-by-line using the `diff-sequences` package
- **format** the changed or common lines using the `chalk` package

> To use `diffLines` as the function name, write either of the following:

Reviewer 2: ```suggestion
To use this function, write either of the following:
```

> - `const diffLines = require('jest-diff');` in a CommonJS module

Reviewer 2: we'll probably go for named exports in jest 25 (we wanna avoid having to use `export =`), but this is of course fine for now

- `import diffLines from 'jest-diff';` in an ECMAScript module

### Example of default export

```js
const a = ['delete', 'change from', 'common'];
const b = ['change to', 'insert', 'common'];

const difference = diffLines(a, b);
```

The returned **string** consists of:

- annotation lines which describe the change symbols with labels
- blank line
- comparison lines in which `Expected` lines are green, `Received` lines are red, and common lines are dim

```diff

> - Expected

Reviewer 2: We should have an option for removing these two lines btw, so it's more useful outside of jest

Author: Yes, I agree. I think that a prerequisite for default export is solve its edge cases. The presence of annotation line is most reliable clue I know that the return string is a diff.

// diffs[0][0] === DIFF_EQUAL
// diffs[0][1] === 'change '

// diffs[1][0] === DIFF_DELETE
// diffs[1][1] === 'fr'

// diffs[2][0] === DIFF_INSERT
// diffs[2][1] === 't'

// Here is a small coincidental common substring:
// diffs[3][0] === DIFF_EQUAL
// diffs[3][1] === 'o'

// diffs[4][0] === DIFF_DELETE
// diffs[4][1] === 'm'
```

## Options

The default options are for the report when an assertion fails from the `expect` package used by Jest.

For other applications, you can provide an options object as a third argument:

- `diffLinesUnified(a, b, options)`
- `diffStringsUnified(a, b, options)`

### Properties of options object

| name                  | default       |
| :-------------------- | :------------ |
| `aAnnotation`         | `'Expected'`  |
| `aColor`              | `chalk.green` |
| `aSymbol`             | `'-'`         |
| `bAnnotation`         | `'Received'`  |
| `bColor`              | `chalk.red`   |
| `bSymbol`             | `'+'`         |
| `commonColor`         | `chalk.dim`   |
| `commonSymbol`        | `' '`         |
| `contextLines`        | `5`           |
| `expand`              | `true`        |

> | `omitAnnotationLines` | `false`       |

Reviewer 2: have an example with this?
